server/lib/404.js

```diff
@@ -8,7 +8,8 @@ module.exports = function (req, res, next) {
   res.status(404);
 
   if (req.accepts('html')) {
-    return res.render('404');
+    res.setHeader('content-type', 'text/html; charset=utf-8');
```

Comment:
I think the problem is the CSP headers are set after the page is rendered, which IIUC essentially resets the render. So, we might have to set the `content-type` header, then call the CSP middleware, then do the render.